Userland/Libraries/LibGfx/ImageFormats/ILBMLoader.cpp

```diff
@@ -233,7 +250,7 @@ static ErrorOr<ByteBuffer> uncompress_byte_run(ReadonlyBytes data, ILBMLoadingCo
 
     u32 index = 0;
     u32 read_bytes = 0;
-    while (read_bytes < length) {
+    while (read_bytes < length - 1) {
```

Comment:
That seems quite unrelated to the purpose of this commit. If this is a bugfix, please make it in its own commit.

Reply:
It is a bugfix and unrelated indeed but I stumbled upon it while testing various iff-24 files.

Do you mean an own commit + PR or another commit in this PR?

Reply:
Separate commit in the same PR is fine.

Reply:
I created another commit for the fix.

Reply:
I have a hard-time understanding what is the fundamental issue.
Because we are not reading past the end of the `Bytes`, as we would hit a `VERIFY` when reading, and I would be surprised that the format *always* include an unneeded byte.

Have you tried images with an odd size? All of our test images have an even number of pixels.

Reply:
Here is an example [file](https://aminet.net/pix/trace/bwing24.lha) (need to uncompress with something supporting lha like `7zip` or `the unarchiver` on mac) that breaks without the patch: it breaks on line 269.

Reply:
I don't doubt that it allows us to open more images, but I am concerned for regressions.
Because as I said earlier, you're not reading one-byte after then end, a `VERIFY` would have caught you.
So it means that you're stopping one byte earlier every single time, what information contains this last byte? If that's a filler byte, when is it included? Can we explicitly write that we are skipping it/them?

Reply:
I understand what you mean but am not sure why such patch is needed. From what I could read about the `byterun1` (also called `PackBits` on Mac) compression algorithm:

> There is no way based on the PackBits data to determine the end of the data stream; that is to say, one must already know the size of the compressed or uncompressed data before reading a PackBits data stream to know where it ends.

On `ffmpeg` they end the decoding loop when either the (packed) buffer has been read or the output buffer is filled, see [here](https://ffmpeg.org/doxygen/0.7/libavcodec_2iff_8c-source.html#l00403).

I updated the loop to be:

```cpp
    while (read_bytes < length && index < plane_data_size) {
    // ...
    }
```

And it also fixes the problem and seems more clear to me.

What do you think?

Reply:
This conversation doesn't seem to be resolved?

Reply:
>also called `PackBits` on Mac

That's why it looked familiar 😅, I also implemented it for the TIFF decoder. We should make a common implementation one you port the decoder to `Stream`.
https://github.com/SerenityOS/serenity/blob/b27a62488c7fcf6bc39dfd927388f575192d1565/Userland/Libraries/LibGfx/ImageFormats/TIFFLoader.cpp#L199-L234

That being said, the issue is probably a bug in the encoder, as a single byte doesn't make sense in a PackBits Stream. Maybe a byte is included to keep the alignment on word boundary 🤷‍♂️.
Anyway, your new proposal is much better IMO as it is both more clear and more robust. You just didn't commit it yet 😄

Reply:
Yes: I wanted your opinion before :) I just pushed the update (I amended the first commit).